### DISCUSSION TITLE
Interrupted signing without signer certificate from start

### DISCUSSION THREAD

**Discussion by naglalakk** [MatthiasValvekens/pyHanko#453] [Q&A]
Hi. I'm following the documentation for an interrupted signing where signer's certificate is not known until a response is received from the signing service. I can't for the life of me figure out what I'm doing wrong. I've studied the structures that I'm suppose to be passing in here and in the end I used the External signer which includes packaging a lot of this data (ContentInfo, SignedData etc). I always get the same results

```
Trust anchor: "No path to trust anchor found."
The signer's certificate is untrusted.


Integrity
---------
The signature is cryptographically unsound.

The digest algorithm used was 'sha256'.
The signature mechanism used was 'sha256_rsa'
```

# Preparing the document

```
async def prepare_document(w):
    pdf_signer = signers.PdfSigner(
        # Specifying a digest algorithm (or signature mechanism)
        # is necessary if the signing cert is not available
        signers.PdfSignatureMetadata(
            field_name="Signature",
            md_algorithm="sha256",
        ),
        signer=signers.ExternalSigner(
            # note the 'None's
            signing_cert=None,
            cert_registry=None,
            signature_value=256,
            signature_mechanism=SignedDigestAlgorithm({"algorithm": "sha256_rsa"}),
        ),
    )

    prep_digest, tbs_document, output = await pdf_signer.async_digest_doc_for_signing(
        w, bytes_reserved=BYTES_RESERVED
    )
    return prep_digest, tbs_document, output
```

# Calling the signing service

```
async def main():
    with open("main/pdf_test/dummy.pdf", "rb") as doc:
        digest_b64 = None
        writer = IncrementalPdfFileWriter(doc)
        prep_digest, tbs_document, output = await prepare_document(writer)

    digest = hashlib.sha256(prep_digest.document_digest).digest()
    digest_b64 = base64.b64encode(digest)
    data_str = digest_b64.decode("utf-8")  # This is sent to the signing service
    resp = client.sign_sim("6906649", "Hallo", data_str)
    raw_signature = resp.id_token_decoded.get("signature")
    signature = base64.b64decode(raw_signature)
    raw_cert = resp.id_token_decoded.get("certificate")
    certificate = base64.b64decode(raw_cert)
    certificate_obj = x509.Certificate.load(certificate)

    await finish_signing(
        output,
        prep_digest,
        signature,
        certificate_obj,
    )
```

# finish signing

```
async def finish_signing(output, prep_digest, signature, certificate):
    digest = hashlib.sha256(prep_digest.document_digest).digest()
    signer = ExternalSigner(
        signing_cert=certificate,
        signature_value=signature,
        signature_mechanism=SignedDigestAlgorithm({"algorithm": "sha256_rsa"}),
        cert_registry=cert_registry,
    )
    signed_attrs = await signer.signed_attrs(
        digest, "sha256", use_pades=False, timestamper=timestamper
    )
    signer_info = signer.signer_info("sha256", signed_attrs, signature)
    content_info = await signer.async_sign_prescribed_attributes(
        "sha256", signed_attrs, timestamper=timestamper
    )
    cert_anchor = CertTrustAnchor(islandsrot_cert)
    audkenni_anchor = CertTrustAnchor(audkenni_cert)
    time_cert_anchor = CertTrustAnchor(tsa_root)
    tsa_path = ValidationPath(
        trust_anchor=time_cert_anchor, interm=[], leaf=tsa_im_cert
    )
    cert_path = ValidationPath(trust_anchor=cert_anchor, interm=[], leaf=audkenni_cert)
    signer_path = ValidationPath(
        trust_anchor=cert_anchor, interm=[audkenni_cert], leaf=certificate
    )
    psv = PreSignValidationStatus(
        signer_path=signer_path,
        validation_paths=[cert_path],
        ts_validation_paths=[tsa_path],
    )
    psi = PostSignInstructions(
        validation_info=psv,
        timestamper=timestamper,
        timestamp_md_algorithm="sha384",
    )
    await PdfTBSDocument.async_finish_signing(
        output,
        prep_digest,
        content_info,
        post_sign_instr=psi,
        validation_context=validation_context,
    )

    with open("main/pdf_test/dummy_signed.pdf", "wb") as signed_doc:
        signed_doc.write(output.getbuffer())
```

My validation context

```
validation_context = ValidationContext(
    trust_roots=[islandsrot_cert],
    other_certs=[audkenni_cert, tsa_im_cert],
    extra_trust_roots=[tsa_root],
    allow_fetching=True,
    revocation_mode="hard-fail",
)
```

Btw, this validator: https://ec.europa.eu/digital-building-blocks/DSS/webapp-demo/validation
tells me that my signature is in PCKS7-T format, but I'm quite sure it's not. It should be pcks1 according to the docs of the signing service.
 
When doing

```
def decode_signature(signature, public_key):

    # Decrypt the signature using the public key
    try:
        decrypted_data = public_key.recover_data_from_signature(
            signature, padding.PKCS1v15(), hashes.SHA256()
        )
        return decrypted_data
    except Exception as e:
        print(f"Failed to decrypt signature: {e}")
        return None
```

I can extract the hash sent to the signing service with the signer certificate. I've also validated that the chain from the signer certificate and up to the root is valid and that returns Ok. 

Eventually I have to use PADES-B-LTA but for now I'm just trying to validate the signature. 
Is there anything you notice here which is off in terms of an interrupted signing process where the certificate is not known?

**Comment by MatthiasValvekens** [comment]
>  tells me that my signature is in PCKS7-T format, but I'm quite sure it's not. It should be pcks1 according to the docs of the signing service.

Don't worry, PKCS#1 and PKCS#7 describe two very different parts of the public-key ecosystem, so it's quite likely that both are correct. PKCS#1 describes (among other things) a way of serialising RSA signatures (low-level), whereas PKCS#7 is a container standard (the predecessor to the cryptographic message syntax (CMS), the two are sometimes used interchangeably).

Before I attempt to parse your code: it's not totally clear from your explanation what is going wrong, actually. What concrete issue is the validator reporting? Can you share a piece of sample output?

**Reply by naglalakk** [reply]
The main issue the validator reports is that the signature is not cryptographically sound and that signer's certificate is not trusted, no path to trust anchor found. Here is the message from validation

```
Trust anchor: "No path to trust anchor found."
The signer's certificate is untrusted.


Integrity
---------
The signature is cryptographically unsound.

The digest algorithm used was 'sha256'.
The signature mechanism used was 'sha256_rsa'
```


This is consistent with this validator:
https://ec.europa.eu/digital-building-blocks/DSS/webapp-demo/validation

which tells me that the signature is not valid and  "The reference data object is not intact!". DSS validator also reports these as failing

* Is the result of the Basic Validation Process acceptable? - FAIL
* Is the result of the LTV validation process acceptable? - FAIL

This is expected though because atm I'm not doing anything to adhere to this (however if I add a separate timestamp for LTV pyHanko stops giving the error message for ltv signatures but the DSS validator result stays the same)

but the DSS validator doesn't seem to have the same problem with the certificates. It displays the chain correctly from signer certificate to root. But both seem to have an issue with the signature. 

I think this might not be correct

```
    digest = hashlib.sha256(prep_digest.document_digest).digest()
    signer = ExternalSigner(
        signing_cert=certificate,
        signature_value=signature,
        signature_mechanism=SignedDigestAlgorithm({"algorithm": "sha256_rsa"}),
        cert_registry=cert_registry,
    )
    signed_attrs = await signer.signed_attrs(
        digest, "sha256", use_pades=False, timestamper=timestamper
    )
    signer_info = signer.signer_info("sha256", signed_attrs, signature)
    content_info = await signer.async_sign_prescribed_attributes(
        "sha256", signed_attrs, timestamper=timestamper
    )
 ```
 
 but I'm trying to tell it, this document content is the signed attribute that was sent to the signing service. However I don't send it as a hashed signed_attrs, I just send it like so to the signing service

```
    digest = hashlib.sha256(prep_digest.document_digest).digest()
    digest_b64 = base64.b64encode(digest)
    data_str = digest_b64.decode("utf-8")  # This is sent to the signing service
    resp = client.sign_sim("6906649", "Hallo", data_str)
```

Main problem is, I cannot get the signature to verify and I'm unsure of how to prepare the document properly if I don't have the signer's certificate from the start. There always seems to be a mismatch for me so I'm wondering which step I'm doing wrong.
